### ISSUE TITLE
`brew install opencv` fails: libIlmImf-2_3.so.24: undefined reference to ... @GLIBCXX_3.4.26

### ISSUE THREAD

**Issue by timotheecour** [Homebrew/linuxbrew-core#21141]
# Bug report

**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?

https://gist.github.com/ef1b3c3e2bddec8d1a6f5baa219bc23d

- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

## What you were trying to do (and why)

`brew install opencv`

## What happened (include command output)

I have to keep running it to circumvent https://github.com/Homebrew/linuxbrew-core/issues/21139; eventually all the other packages get installed and then I get a new error when installing opencv

you won't get any useful error in terminal, you have to look at gist-logs output https://gist.github.com/ef1b3c3e2bddec8d1a6f5baa219bc23d

<details>
  <summary>Command output</summary>
  <pre>

https://gist.githubusercontent.com/timotheecour/ef1b3c3e2bddec8d1a6f5baa219bc23d/raw/b12db563957b481ddf1228cb2744084728761403/02.make
/home/linuxbrew/.linuxbrew/bin/ld: /lib/x86_64-linux-gnu/libIlmImf-2_3.so.24: undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()@GLIBCXX_3.4.26'
/home/linuxbrew/.linuxbrew/bin/ld: /lib/x86_64-linux-gnu/libIlmThread-2_3.so.24: undefined reference to `typeinfo for std::thread::_State@GLIBCXX_3.4.22'
/home/linuxbrew/.linuxbrew/bin/ld: /lib/x86_64-linux-gnu/libIlmThread-2_3.so.24: undefined reference to `std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)())@GLIBCXX_3.4.22'
/home/linuxbrew/.linuxbrew/bin/ld: /lib/x86_64-linux-gnu/libIlmThread-2_3.so.24: undefined reference to `std::thread::_State::~_State()@GLIBCXX_3.4.22'
collect2: error: ld returned 1 exit status
make[2]: *** [apps/annotation/CMakeFiles/opencv_annotation.dir/build.make:113: bin/opencv_annotation] Error 1
make[2]: Leaving directory '/tmp/opencv-20200925-193732-nfmpp9/opencv-4.4.0/build'
make[1]: *** [CMakeFiles/Makefile2:5684: apps/annotation/CMakeFiles/opencv_annotation.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

  </pre>
</details>

## What you expected to happen

works

## Step-by-step reproduction instructions (by running `brew install` commands)

`brew install opencv` (keep re-running to circumvent https://github.com/Homebrew/linuxbrew-core/issues/21139)

**Comment by danielnachun** [comment]
It appears to me that this may be caused by the same issue I seem to encounter when I try to build opencv: cmake stubbornly insists on checking if libraries are available in "default" locations like ```/lib``` **before** checking the linuxbrew root.  It will pick up Linuxbrew libraries if the library isn't available on the system, but if it is, it will use the system library instead.  If this library is older than what is in Linuxbrew, it leads to a conflict.  For you it seems it is because it is using the system version of ```ilmbase``` which is linked to an older version of ```libstdc++``` than what Linuxbrew ships with.  For me it is ```libtiff```.  I have looked for a way to stop this behavior but have not been successful in doing so yet.

**Comment by danielnachun** [comment]
I was able to get ```opencv``` to compile on my system by adding the following parameters to the CMake arguments to "hide" system libraries:

```
    args = std_cmake_args + %W[
      -DTIFF_LIBRARY=#{Formula["libtiff"].opt_prefix}/lib/libtiff.so
      -DZLIB_LIBRARY=#{Formula["zlib"].opt_prefix}/lib/libz.so
      -DJPEG_LIBRARY=#{Formula["libjpeg"].opt_prefix}/lib/libjpeg.so
      -DPNG_LIBRARY=#{Formula["libpng"].opt_prefix}/lib/libpng.so
      -DPROTOBUF_LIBRARY=#{Formula["protobuf"].opt_prefix}/lib/libprotobuf.so
      -DOpenBLAS_LIB=#{Formula["openblas"].opt_prefix}/lib/libopenblas.so
      -DCMAKE_OSX_DEPLOYMENT_TARGET=
      -DBUILD_JASPER=OFF
      -DBUILD_JPEG=OFF
      -DBUILD_OPENEXR=OFF
      -DBUILD_PERF_TESTS=OFF
      -DBUILD_PNG=OFF
      -DBUILD_PROTOBUF=OFF
      -DBUILD_TESTS=OFF
      -DBUILD_TIFF=OFF
      -DBUILD_WEBP=OFF
      -DBUILD_ZLIB=OFF
      -DBUILD_opencv_hdf=OFF
      -DBUILD_opencv_java=OFF
      -DBUILD_opencv_text=ON
      -DOPENCV_ENABLE_NONFREE=ON
      -DOPENCV_EXTRA_MODULES_PATH=#{buildpath}/opencv_contrib/modules
      -DOPENCV_GENERATE_PKGCONFIG=ON
      -DPROTOBUF_UPDATE_FILES=ON
      -DWITH_1394=OFF
      -DWITH_CUDA=OFF
      -DWITH_EIGEN=ON
      -DWITH_FFMPEG=ON
      -DWITH_GPHOTO2=OFF
      -DWITH_GSTREAMER=OFF
      -DWITH_JASPER=OFF
      -DWITH_OPENEXR=ON
      -DWITH_OPENGL=OFF
      -DWITH_QT=OFF
      -DWITH_TBB=ON
      -DWITH_V4L=OFF
      -DWITH_VTK=ON
      -DBUILD_opencv_python2=OFF
      -DBUILD_opencv_python3=ON
      -DPYTHON3_EXECUTABLE=#{Formula["python@3.9"].opt_bin}/python3
    ]
```

To solve the original issue, I believe it would also be necessary to have these additional CMake arguments:

```      
      -DOPENEXR_ILMIMF_LIBRARY=#{Formula["openexr"].opt_prefix}/lib64/libIlmImf.so
      -DOPENEXR_ILMTHREAD_LIBRARY=#{Formula["ilmbase"].opt_prefix}/lib/libIlmThread.so
```

So for anyone having these kinds of linking errors, the solution is to "hide" the system libraries by manually specifying them.  There may be a more clever way to solve this, but at least this works!

**Comment by danielnachun** [comment]
I should also note that I had to fix an issue with ```libpng``` putting its header files in ```$(brew --prefix)/include/libpng16``` instead of ```$(brew --prefix)/include/libpng```, which made it impossible for ```opencv``` to find them.  I also had to disable V4L support because that isn't available as a Homebrew package I could find.  Lastly, I also had to fix both ```ilmbase``` and ```openexr``` to install to ```lib``` instead of ```lib64```.  I will try to make PRs for these fixes when I get a chance.
